Option for controller-specific duration parameters

2scholz opened the pull request:
`allowed_execution_duration_scaling` and `allowed_goal_duration_margin` were previously set globally and could not be specified for individual controllers.
This patch makes it possible to define the values for controllers in the `controller_list` parameter (they can be specified in the `controllers.yaml`).
A short example: 
``` 
controller_list:
 - name: example_controller
   type: FollowJointTrajectory
   allowed_execution_duration_scaling: 1.2
   allowed_goal_duration_margin: 0.5
```
The global value is still used whenever no controller-specific value was defined. 

@v4hn

v4hn commented:
I supervised the development of this patch.

These values are controller-specific to begin with, but nobody complained before that they could not be set for individual controllers...
In our use-case the execution time of our robotic gripper can deviate quite a lot from the planned trajectory, but the robotic arm it is attached to is very reliable.

Should not be cherry-picked because it breaks ABI.

Please review @davetcoleman @rhaschke (sorry Robert, I know you're busy :))

davetcoleman commented on moveit_ros/planning/trajectory_execution_manager/include/moveit/trajectory_execution_manager/trajectory_execution_manager.h:
```diff
@@ -343,6 +343,10 @@ class TrajectoryExecutionManager
   bool execution_duration_monitoring_;
   double allowed_execution_duration_scaling_;
```
add comment above that these are the 'global' variables, in contrast to the controller specific ones

davetcoleman commented on moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp:
```diff
@@ -129,6 +129,25 @@ void TrajectoryExecutionManager::initialize()
   else
     allowed_goal_duration_margin_ = DEFAULT_CONTROLLER_GOAL_DURATION_MARGIN;
 
+  XmlRpc::XmlRpcValue controller_list;
```
add comment above saying what this section does

also lets just move this into a private function ``loadControllerParams()``

davetcoleman commented on moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp:
```diff
@@ -1421,12 +1440,20 @@ bool TrajectoryExecutionManager::executePart(std::size_t part_index)
+      // expected duration is the duration of the longest part
+      expected_trajectory_duration =
+          std::max(d * current_scaling + ros::Duration(current_margin), expected_trajectory_duration);
+    }
```
awesome!

rhaschke commented on moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp:
```diff
@@ -1421,12 +1440,20 @@ bool TrajectoryExecutionManager::executePart(std::size_t part_index)
-    expected_trajectory_duration = expected_trajectory_duration * allowed_execution_duration_scaling_ +
-                                   ros::Duration(allowed_goal_duration_margin_);
+      // use controller-specific values if defined
+      const double current_scaling = controller_allowed_execution_duration_scaling_.count(context.controllers_[i]) > 0 ?
```
This is inefficient, as the map is searched two times: once for count() and once for access.
Better use find() to search once and subsequently compare the retrieved iterator against end().

2scholz commented:
Thanks for the feedback. I added the requested changes to the commit and made a pull request to add the option in the tutorials (see https://github.com/ros-planning/moveit_tutorials/pull/161)

2scholz commented:
The parameters handled by `dynamic reconfigure` need to be declared in the `.cfg` file. 
There is an old open [issue](https://github.com/ros/dynamic_reconfigure/issues/62) where the possibility to add parameters at runtime was discussed, but nothing came of it yet.

rhaschke commented:
@2scholz Could you please rebase? Also, discuss/document the relationship of the controller-specific variables and their dynamically reconfigurable global counterparts.
I guess, controller-specific variables override any globals.

2scholz commented:
I rebased the branch.
I changed the comment in the header file above the controller-specific variables a bit to make it clearer that these override the global ones. 
Additionally a description of the difference between the global and controller-specific values was added in the pull request for the documentation: https://github.com/ros-planning/moveit_tutorials/pull/161

rhaschke commented on moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp:
```diff
@@ -1415,12 +1418,24 @@ bool TrajectoryExecutionManager::executePart(std::size_t part_index)
 
-    expected_trajectory_duration = expected_trajectory_duration * allowed_execution_duration_scaling_ +
-                                   ros::Duration(allowed_goal_duration_margin_);
+      // use controller-specific values if defined
```
Please add a comment here, that the global parameters are dynamically reconfigurable (and thus might change from call to call). If there are controller-specific values defined, they override any global parameters - even if set through dynamic reconfigure.

That's the point, I'm feeling still uncomfortable with. If the user has changed the global parameters, they will not take effect and the user isn't notified about that fact...
Did you considered to use the minimum of global and specific?

v4hn replied:
@rhaschke We were discussing over here where exactly you wanted to have more documentation and @2scholz added more explanation to the tutorial already: https://github.com/ros-planning/moveit_tutorials/pull/161 .

I see your concerns about dynamic_reconfigure not working, but the only way to get around this is to patch dynamic_reconfigure and allow runtime variables in there. I don't like that very much.

Neither minimum nor maximum operator make sense in practice.

rhaschke replied:
I think, there should be a reminder in the source code (at indicated line 1422), that controller-specific values will override dynamically configurable globals. I don't yet have a solution for this as well, but there should be a reminder in the source code if we come up with a solution later...

Eventually there should also be a warning (once) if global parameters have changed through dynamic reconfigure, but are overridden by controller-specific ones. All you need to do for this, is to keep track of the dyn-reconfig-callbacks (I think there is always one to set the initial value).

v4hn replied:
I just added a comment there.
